File: Belajar Android/RecycleViewCardView/app/src/main/java/com/komputerkit/recycleviewcardview/MainActivity.java
package com.komputerkit.recycleviewcardview;

import android.os.Bundle;
import android.view.View;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    RecyclerView rvDataSiswa;
    SiswaAdapter siswaAdapter;
    List<Siswa> daftarSiswa;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        inisialisasiRecyclerView();
        isiDataAwal();
    }

    public void inisialisasiRecyclerView(){
        rvDataSiswa = findViewById(R.id.rcvSiswa);
        rvDataSiswa.setLayoutManager(new LinearLayoutManager(this));
    }

    public void isiDataAwal(){
        daftarSiswa = new ArrayList<>();
        daftarSiswa.add(new Siswa("Alif Ramadhan", "Bandung"));
        daftarSiswa.add(new Siswa("Dewi Lestari", "Semarang"));
        daftarSiswa.add(new Siswa("Andika Pratama", "Medan"));
        daftarSiswa.add(new Siswa("Citra Kirana", "Makassar"));
        daftarSiswa.add(new Siswa("Reza Mahendra", "Palembang"));
        daftarSiswa.add(new Siswa("Laila Amalia", "Bogor"));
        daftarSiswa.add(new Siswa("Fauzan Hakim", "Malang"));

        siswaAdapter = new SiswaAdapter(this, daftarSiswa);
        rvDataSiswa.setAdapter(siswaAdapter);
    }

    public void btnTambah(View view) {
        daftarSiswa.add(new Siswa("Nadya Zahra", "Solo"));
        siswaAdapter.notifyDataSetChanged();
        rvDataSiswa.scrollToPosition(daftarSiswa.size() - 1);
    }
}

File: 3. Android Studio Firebase/TurtorialFireBase/app/src/main/java/com/komputerkit/firebasetutorial/MainActivity.kt
package com.komputerkit.firebasetutorial

import android.app.NotificationManager
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.Firebase
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.analytics
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.FirebaseMessagingService
import java.util.jar.Manifest
import kotlin.math.log

@RequiresApi(Build.VERSION_CODES.O)
class MainActivity : AppCompatActivity() {
    private lateinit var firebaseAnalytics: FirebaseAnalytics

    companion object {
        const val CHANNEL_ID = "dummy_channel";
    }

    private lateinit var requestPermissionLauncher: ActivityResultLauncher<String>
    private val notifiManager: NotificationManager by lazy {
        getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        firebaseAnalytics = Firebase.analytics
        requestPermissionLauncher =registerForActivityResult(ActivityResultContracts.RequestPermission()) {
            if (it) {

            }else {
                Snackbar.make(
                    findViewById<View>(android.R.id.content).rootView,
                    "Please grant Notification permission from app Settings",
                    Snackbar.LENGTH_LONG
                ).show()
            }
        }
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        val crashButton = Button(this)
        crashButton.text = "Test Crash"
        crashButton.setOnClickListener {
            throw RuntimeException("Test Crash") // Force a crash
        }

        addContentView(
            crashButton, ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        )

        requestPermission()
        createToken();
        createNotificationChannel()
    }

    private fun requestPermission () {
        if (ContextCompat.checkSelfPermission(
                this,
                android.Manifest.permission.POST_NOTIFICATIONS,
            ) == PackageManager.PERMISSION_GRANTED
        ) {

        } else {
            requestPermissionLauncher.launch(android.Manifest.permission.POST_NOTIFICATIONS)
        }
    }

    private fun createToken() {
        val TAG = "FCM__TOKEN"

        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            if (!task.isSuccessful) {
                Log.w(TAG, "Fetching FCM registration token failed", task.exception)
                return@addOnCompleteListener
            }

            val token = task.result
            Log.d(TAG, token)
        }
    }

    private fun createNotificationChannel() {
        val name = "Important Notification Channel"
        val descriptionText = "This notification contains important announcements, etc."
        val importance = NotificationManager.IMPORTANCE_HIGH
        val channel = android.app.NotificationChannel(CHANNEL_ID, name, importance).apply {
            description = descriptionText
        }

        notifiManager.createNotificationChannel(channel)
    }

}
